Spark: Adding simple custom partition sort order option to RewriteManifests Spark Action

zachdisc opened the pull request:
## What
This adds a simple `sort` method to the `RewriteManifests` spark action which lets user specify the partition column order to consider when grouping manifests. 

Illustration: 

```
RewriteManifests.Result result =
        actions
            .rewriteManifests(table)
            .sort("c", "b", "a")  < -- this is the new api piece
            .execute();
```

Closes https://github.com/apache/iceberg/issues/9615


## Why
Iceberg's metadata is organized into a forest of manifest_files which point to data files sharing common partitions. By default, and during `RewriteManifests`, the partition grouping is determined by the default `Spec` partition order. If the primary query pattern is more aligned with the last partition in the table's spec, manifests are poorly suited to quickly plan and prune around those partitions. 

EG 
```
CREATE TABLE
...
PARTITIONED BY (region, storeId, bucket(ipAddress, 100), days(event_time)
```
Will create manifests that first group by `region`, whose `manifest_file` contents may span a wide range of `event_time` values. For a primary query pattern that doesn't care about `region`, `storeId`, etc, this leads to inefficient queries. 



## Requested Feedback and decisions
* I chose to make the input to `sort` be the _raw column names_ used in partitioning, not the internal hidden ones. AKA `event_time` instead of `event_time_day`. `foo` instead of `foo_bucket_1234`. Thoughts? Could readily allow both, or just the real, hidden partition column names if people prefer 
* I would next like to have a more capable functional interface `sort(row -> {... return groupingString})`, but was struggling to express a Java `Function`like input that could be used on a `DataSet<Row>`'s `data_file` struct - any pointers on how to parse a `Row`'s struct into a native Pojo and supply a function that can be used like in a UDF here would be apprecaited!

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +45,16 @@ public interface RewriteManifests
+   *
+   * <p>If not set, manifests will be rewritten in the order of the table's spec.
+   *
+   * @param partitionSortOrder a list of partition field names
```
Should this be partition transform names? Considering we will also have multi-arg transforms coming up I think that is probably more accurate.

zachdisc replied:
Ack - didn't know about multi-arg transforms! 

I originally had at as expecting the transform name (`sometime_day` transformed name instead of the underlying `sometime` value). Makes the implementation easier but makes the user have to know what the internally transformed names are referred to as. With multi-arg I think this no longer viable for sure. Could always accept both and check against either the value or transform name. Think that is just too convoluted?

RussellSpitzer replied:
Hmm, I think that's probably safe to do but there will be some edge cases that may be weird. Like i have a column named "identity" :/ 

If we can guard against any ambiguous situations I think that could be fine

jackye1995 replied:
I see the question in the description about 

> I would next like to have a more capable functional interface sort(row -> {... return groupingString}), but was struggling to express a Java Functionlike input that could be used on a DataSet<Row>'s data_file struct - any pointers on how to parse a Row's struct into a native Pojo and supply a function that can be used like in a UDF here would be apprecaited!

I think it aligns with my main comment, which is that we discussed the potential to expose it as a generic function but seems like this PR directly skipped it. Seems like there are some pieces missing, let me see if I can fill it up.

At API level, we could express it as a function using `PartitionData` as the transformation source:

```
default RewriteManifests sort(List<String> partitionFieldNames) {
  return sort(...) // see partitionSpec.partitionToPath for an example of how to convert the partition fields to string
}

RewriteManifests sort(Function<PartitionData, String> partitionFieldsSortStrategy) {
  throw new UnsupportedOperationException();
}

```

and when implementing in Spark, the biggest question is now how do we convert the Row `data_file.partition` to `PartititionData` so we can apply this transform and create the new string column in the data frame that is used for sorting.

I think this is achievable by creating a UDF that transforms the `data_file.partition` Row to `SparkStructLike` using `SparkStructLike.wrap(row)`, and then create a new `PartitionData` based on that, and finally apply the input function to create its string value. You can see an example of that in `PartitionsTable.toPartitionData`. 

@RussellSpitzer @nastra let me know if you agree with this general direction, or if it is becoming too convoluted.

zachdisc replied:
Meta question - since I was struggling to get that part to work, and to keep things more concise,I had intended to submit that additional interface as a separate PR, would it be better to include here in one go? 

Thank you for the pointer there! I'll give this a shot. Converting to something useable in a UDF was what was throwing me for a loop

jackye1995 replied:
My feeling is that if we merge this and you end up having to refactor it a lot, then it is better for everyone to help you make the right changes here. 

But I am all for incremental changes, so if we can agree that the function-based API proposed in the original issue can be added through refactoring later, then I am totally good with that! Let's see what others think.

zachdisc replied:
Sounds good, open to anything. I don't believe there would be much of a refactor at all FWIW. Just a new api option and single branch.

zachdisc replied:
Clarifying on the signature you're suggesting

```
RewriteManifests sort(Function<PartitionData, String> partitionFieldsSortStrategy) {
  throw new UnsupportedOperationException();
}
```

I would propose to have the method operate over a `DataFile`, so that all attributes are available for scrutiny if the user wants, not just the partition information itself. 

```
RewriteManifests sort(Function<DataFile, String> partitionFieldsSortStrategy) {
  throw new UnsupportedOperationException();
}
```

So the specific implementation would be a `SparkDataFile` I think, and need to do the same kind of `toDataFile` converter?

jackye1995 replied:
ah that's actually a great idea! In that case we can actually use the `SparkDataFile` to convert the `data_file` row. You can see an example of that in test like https://github.com/apache/iceberg/blob/main/spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataFile.java#L180-L188

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +45,16 @@ public interface RewriteManifests
   RewriteManifests rewriteIf(Predicate<ManifestFile> predicate);
 
+  /**
+   * Rewrite manifests in a given order, based on partition columns
```
based on partition (values|transforms)

One of these two ^

zachdisc replied:
Great clarification.

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +45,16 @@ public interface RewriteManifests
+  /**
+   * Rewrite manifests in a given order, based on partition columns
+   *
+   * <p>If not set, manifests will be rewritten in the order of the table's spec.
```
"in the order of the transforms in the table's current partition spec"

zachdisc replied:
Ack

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -90,13 +93,45 @@ public class RewriteManifestsSparkAction
+   * @return this action
+   */
+  @Override
+  public RewriteManifestsSparkAction sort(List<String> partitionSortOrder) {
```
As mentioned above in the api spec, this has to be based on transform names and not column names.

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -208,7 +243,15 @@ private RewriteManifests.Result rewriteManifests(ManifestContent content) {
       newManifests = writeUnpartitionedManifests(content, manifestEntryDF, targetNumManifests);
     } else {
-      newManifests = writePartitionedManifests(content, manifestEntryDF, targetNumManifests);
+      if (partitionSortColumns != null) {
```
Should both of these branches be the same? Just one uses the default ordering and one doesn't? That is what we declared in the api above

zachdisc replied:
That makes sense. I was probably just hesitant to touch the main working path with a custom extractor, but I can make that switch.

RussellSpitzer replied:
I think your change here is a great time to eliminate the other path as well

RussellSpitzer commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -84,6 +89,7 @@
 import org.junit.jupiter.api.TestTemplate;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.api.io.TempDir;
+import scala.Tuple2;
```
I would avoid using Scala classes here, we have a Pair utility class you can use in it's place

zachdisc replied:
Great, thanks!

RussellSpitzer commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +472,164 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
   }
 
+  @TestTemplate
+  public void testRewriteManifestsPartitionedTableWithInvalidSortColumnsThowsException()
```
These are very through tests

zachdisc replied:
Too detailed/fragile? I always struggle to come up with simple unit tests for things that involve contextual data layout.

RussellSpitzer replied:
Sorry, I was complementing I think you are covering this well

RussellSpitzer commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +472,164 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+
+    SparkActions actions = SparkActions.get();
+
+    String stagingLocation = java.nio.file.Files.createTempDirectory(temp, "junit").toString();
```
Any reason why we are using another directory here? We could just use our table location (or a sub directory within there)

zachdisc replied:
No real reason - I just took this pattern from another test, will verify that just using the table's location works instead.

zachdisc commented on .palantir/revapi.yml:
```diff
@@ -873,6 +873,10 @@ acceptedBreaks:
       new: "method void org.apache.iceberg.encryption.Ciphers::<init>()"
       justification: "Static utility class - should not have public constructor"
   "1.4.0":
+    org.apache.iceberg:iceberg-api:
```
In this case, I can just add a new `"1.5.0"` section in this config file? Or is there something more involved (forgive the naive question, this is new for me :-) )

zachdisc replied:
Okay Ack, I'll add a default implementation. Should be safe to remove the breaking change piece from revapi.yml then?

zachdisc commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -90,13 +93,45 @@ public class RewriteManifestsSparkAction
+    Map<String, String> partitionFieldMap =
+        spec.fields().stream()
+            .map(
+                partitionField ->
```
> case issues

Mind expanding? Regardless, if I switch to accept just partition transform names then I don' t believe this will be required.

zachdisc commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -90,13 +93,45 @@ public class RewriteManifestsSparkAction
+    // Check if these partitions are included in the spec
+    Preconditions.checkArgument(
+        partitionSortOrder.stream().allMatch(partitionFieldMap::containsKey),
+        "Invalid sorting columns. Expected to match column names used for partitioning in spec: %s."
```
Ack, thanks!

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -90,13 +93,45 @@ public class RewriteManifestsSparkAction
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // Check if these partitions are included in the spec
+    Preconditions.checkArgument(
```
I'm not sure I understand the comment here?

zachdisc replied:
Say there are 2 partition specs for this table. Spec1 partitions by (`a_bucket, b`), Spec2 partitions by (`a_bucket_100, c_days`). If you originally pass a valid `sort(b, a)`, then change the `.specId` to 2, then the validated custom column order will now be inaccurate.

nastra commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +45,16 @@ public interface RewriteManifests
+   * @param partitionSortOrder a list of partition field names
+   * @return this for method chaining
+   */
+  RewriteManifests sort(List<String> partitionSortOrder);
```
yes that should have a default implementation to avoid adding a breaking change. See also https://github.com/apache/iceberg/blob/main/CONTRIBUTING.md#adding-a-default-implementation

zachdisc commented:
## Rev 2
* Default implementation in `RewriteManifests` interface
* Switched to using hidden transform names to specify sort order, not raw column name
* Addressed other PR feedback

Going to work on the Function based api update next. Open to doing so in a new PR or this one.

jackye1995 commented:
Looks like the CI test failed, @zachdisc could you take a look and fix?

Overall, we have the 2 options based on this thread: https://github.com/apache/iceberg/pull/9731#discussion_r1494774386

option 1: we merge this and have a subsequent PR for the functional API and refactor the current change into it
option 2: we directly update the current PR to go with that approach

Is there any preference for which way we go?

zachdisc commented:
> Looks like the CI test failed, @zachdisc could you take a look and fix?
> 
> Overall, we have the 2 options based on this thread: [#9731 (comment)](https://github.com/apache/iceberg/pull/9731#discussion_r1494774386)
> 
> option 1: we merge this and have a subsequent PR for the functional API and refactor the current change into it option 2: we directly update the current PR to go with that approach
> 
> Is there any preference for which way we go?

I'm not quite sure why that failed, I ran spotlessApply and saw a successful build locally before the last push. Regardless, I have two changes inbound, so I'll fix along with the new updates.

zachdisc commented:
## Rev 3
* Fixed styling which caused CI/CD to fail
* Changed the clustering column to be a struct referencing the actual columns instead of a contacted string `p1::p2::p3`.

Thanks to @ Nandhakumar Saravana for pointing this out, when clustering by a string `a::b::c` spark doesn't know about relative priority/hierarchy. AKA `a::b::c` should be close to `a::b::d`, which it does when clustering by actual columns and not a String literal. 

Change imminent for `Function`al interface addition too, just running last tests

jackye1995 commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -90,13 +94,13 @@ public class RewriteManifestsSparkAction
           .rewrittenManifests(ImmutableList.of())
           .addedManifests(ImmutableList.of())
           .build();
-
```
nit: avoid unnecessary newline changes

jackye1995 commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -90,13 +94,13 @@ public class RewriteManifestsSparkAction
   public static final String USE_CACHING = "use-caching";
   public static final boolean USE_CACHING_DEFAULT = false;
 
+  private List<String> partitionSortColumns = null;
```
nit: we follow `public static final` -> `private static final` -> `private final` -> `private` to order variables

zachdisc replied:
Is this a problem? I have these attributes as being set to `null` but able to be redefined. So I think `private` is correct here, its inline with the other attributes in this class.

jackye1995 commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +45,20 @@ public interface RewriteManifests
+   * @param partitionSortOrder a list of partition field transform names
+   * @return this for method chaining
+   */
+  default RewriteManifests sort(List<String> partitionSortOrder) {
```
I think this should be better called `partitionFieldSortOrder`? because the inputs are not the actual partitions, but the partition fields.

zachdisc replied:
Works for me, apologies I'm still struggling with the naming conventions a bit.

jackye1995 commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -160,6 +164,34 @@ public RewriteManifestsSparkAction stagingLocation(String newStagingLocation) {
     return this;
   }
 
+  /**
```
I think we do not need to javadoc the implementation methods. If there is anything we did not mention in the javadoc of the interface, we should add it there.

jackye1995 commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -250,12 +282,40 @@ private List<ManifestFile> writeUnpartitionedManifests(
 
+    // Extract desired clustering/sorting criteria into a dedicated column
+    Dataset<Row> clusteredManifestEntryDF;
+    String clusteringColumnName = "__clustering_column__";
```
nit: this can be a private static variable

jackye1995 replied:
also if this is for sorting, should we call this `__sorting_column__`?

zachdisc replied:
I'm planning to reuse this same column for the more generic `Function` based approach. No strong opinion on naming convention though, or could use different names if you'd rather.

jackye1995 commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -160,6 +164,34 @@ public RewriteManifestsSparkAction stagingLocation(String newStagingLocation) {
+    Set<String> availablePartitionNames =
+        spec.fields().stream().map(PartitionField::name).collect(Collectors.toSet());
+
+    // Check if these partitions are included in the spec
```
```suggestion
    // Check if these partition fields are included in the spec
```

jackye1995 commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -250,12 +282,40 @@ private List<ManifestFile> writeUnpartitionedManifests(
-          Dataset<Row> transformedDF = repartitionAndSort(df, partitionColumn, numManifests);
+          Column partitionColumn = df.col(clusteringColumnName);
+          Dataset<Row> transformedDF =
+              repartitionAndSort(df, partitionColumn, numManifests).drop(clusteringColumnName);
```
The column here is more used for repartitioning, rather than sorting. Sorting only happens for rows with the same clustering column value. I am starting to think if we should call the API `repartition(List<String> partitionFields)`, rather than `sort(List<String> partitionFields)`.

zachdisc replied:
Happy to make that change, but IMHO the API is indicating how the entire forest of manifests is sorted, not necessarily how each tree is sorted. Like, its sorting the `manifest_list` entries in the end, while the `manifest_files` are sorted further within that. Thoughts?

zachdisc replied:
@nastra @jackye1995 thoughts on a general rename to `cluster()` for this api action?

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+                    .sort(badSortKeys1)
+                    .execute())
+        .isInstanceOf(IllegalArgumentException.class)
+        .message()
```
this should just use `.hasMessageContaining` instead of `.message().contains()`

zachdisc replied:
Good call, this caught a bug with my error message too!

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+                    .execute())
+        .isInstanceOf(IllegalArgumentException.class)
+        .message()
+        .contains("Cannot use custom sort order");
```
same as above

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+    // So the upper bound of the partitions tracked by the first file should be LEQ the lower bounds
+    // of the second. Etc
+    assertThat(uppers.get(0))
+        .withFailMessage(
```
I think what you rather want here is using `.as(...)` because it will then show actual/expected values, whereas `.withFailMessage()` will only show this message without actual/expected

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+    c3Boundaries.forEach(
+        boundary -> {
+          assertThat(boundary.second() - boundary.first())
+              .withFailMessage(
```
same as above

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+    // manifest should
+    // include the lower bound of 0, and the last should have the upper bound of 9
+    assertThat(lowers.get(0))
+        .withFailMessage("Lower bound of first manifest partition should be 0")
```
same as above

nastra replied:
unresolving this comment since it hasn't been updated yet

zachdisc replied:
My bad, missed these. Fixed all occurrences.

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+
+    // Rewritten manifests are clustered by c3_bucket - each should contain only a subset of the
+    // lower and upper bounds
+    // of the partition 'c3'.
```
nit: can go into the line above

zachdisc replied:
Ugh. The spotlessApply build actions are great but I miss when they break things like this.

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
   }
 
+  @TestTemplate
+  public void testRewriteManifestsPartitionedTableWithInvalidSortColumnsThowsException()
```
```suggestion
  public void testRewriteManifestsPartitionedTableWithInvalidSortColumns()
```

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).identity("c1").bucket("c3", 10).build();
+    Map<String, String> options = Maps.newHashMap();
+    options.put(TableProperties.FORMAT_VERSION, String.valueOf(formatVersion));
+    options.put(TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED, snapshotIdInheritanceEnabled);
```
are these options relevant for testing this behavior?

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+        .message()
+        .contains("Cannot use custom sort order");
+
+    // c3_bucket is the correct internal partition name to use, c3 is the untransformed colum name,
```
typo: colum -> column

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +471,163 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+
+    // c3_bucket is the correct internal partition name to use, c3 is the untransformed colum name,
+    // sort() expects
+    // the hidden partition column names
```
nit: can go into the live above this

zachdisc commented:
## Rev 3
* Addressed last round of comments
* Added a new API option to sort with a custom supplied `Function<DataFile>`. 

Illustration (in unit test) - say you have a table that is partitioned by `a, truncate(b, 5), bucket(10, c)`. We can choose to re-organize manifests around `c`, such that we have roughly half of the `c` partitions contained in individual `manifest_list` tree entry. 

So supply a Function like below, which checks if the `c` partition value is < 5 and returns a string. 

```
Function<DataFile, String> test =
        (Function<DataFile, String> & Serializable)
            (dataFile) -> {
              StructLike partition = dataFile.partition();
              // Find the ordinal index for the c3 partition column for this data file
              int c3Index =
                  IntStream.range(0, spec.fields().size())
                      .filter(i -> spec.fields().get(i).name().contains("c3"))
                      .findFirst()
                      .getAsInt();
              Object c3BucketValue = partition.get(c3Index, Object.class);

              // Return one string for the lower values, one for the upper. RewriteManifests
              // will cluster datafiles together in manifests according to this value.
              return (Integer) c3BucketValue < 5 ? "cluster=LT_5" : "cluster=GTE_5";
            };
```

The value used for repartitioning/sorting on will fall into two values. One where the transformed bucket value is less than 5, one for values over 5. Illustration of the partition set for a data file vs this new supplied clustering column value


```
+--------------------+---------------------+
|partition           |__clustering_column__|
+--------------------+---------------------+
|{0, -531806488, 0}  |cluster=LT_5         |.  <-- the last partition value for `c` is 0. Which is labeled as "LT_5" 
|{1, 385955472, 7}   |cluster=GTE_5        |
|{2, 604077840, 6}   |cluster=GTE_5        |
|{3, 1875302972, 4}  |cluster=LT_5         |
|{4, -1772544904, 0} |cluster=LT_5         |
+--------------------+---------------------+
```

Rewritten manifests will group manifest_file s by having data files within that map to `c` partition values from 0-4 and 5-9

nastra commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +47,38 @@ public interface RewriteManifests
+   * @return this method for chaining
+   */
+  default RewriteManifests sort(Function<DataFile, String> sortStrategyFunction) {
+    throw new UnsupportedOperationException();
```
this needs a msg

zachdisc replied:
Ack, my bad

jackye1995 replied:
I think this is not resolved yet?

zachdisc replied:
I haven't pushed this commit yet. Will shortly

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +474,309 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+        .withFailMessage("Lower bound of second manifest c3 bucket partition should be 5")
+        .isEqualTo(5);
+    assertThat(uppers.get(1))
+        .withFailMessage("Upper bound of second manifest c3 bucket partition should be 9")
```
all of these should use `.as()` instead of `.withFailMessage()`

zachdisc replied:
Ack

zachdisc commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +474,309 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+    // Let's say I want to sort manifests into buckets 0-4 and 5 - 9
+    // I know my Table and partition Spec, so can do that programmatically
+    Function<DataFile, String> test =
+        (Function<DataFile, String> & Serializable)
```
I could not figure out a way to pass a `Function` around for use in UDFs without declaring it like this at the outset. Looking for ideas here, this part of the API is a bit clunky.

jackye1995 commented:
@zachdisc Hey Zach, sorry I was trying to experiment with refactoring the current PR, but somehow the branch is public and I ended up directly committing to it...

Let me know if you agree with this or not, overall it tries to express the partition column sort order approach as a kind of sort function to minimize the if-else branches.

jackye1995 commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/CustomDataFileSorterUdf.java:
```diff
@@ -0,0 +1,55 @@
+ * org.apache.iceberg.DataFile} appearing object and pass that along to a runtime supplied Function.
+ * The result of the execution of which must be a String.
+ */
+public class CustomDataFileSorterUdf implements UDF1<Row, String>, Serializable {
```
could we move this as an inner class of the Spark action class? Usually we try to minimize the number of public classes to minimize backwards incompatibility.

zachdisc replied:
I tried but could not get it to work. Kept getting

```
Caused by: java.io.NotSerializableException: org.apache.iceberg.spark.actions.RewriteManifestsSparkAction
```

I attempted some basic `Extends Serializable` mitigations but didn't get anywhere and didn't wish to modify the main class too much. Would you prefer I create a utility package to hold this class at least?

jackye1995 commented:
There is a more general question of why this refactoring cannot be done at API level, the key issue is that `RewriteManifests` cannot access the `Table` object so it cannot construct the sorting string. But I think that is a common problem of all the new actions, we can probably address separately from this PR.

zachdisc commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -628,4 +613,50 @@ private Table table() {
+            field
+                .transform()
+                .toHumanString(fieldType, get(dataFile.partition(), fieldIndex, javaClass));
+        sb.append(fieldValue);
```
I actually had something very much like this ready to go earlier!

The issue with this approach (which was non obvious to me at first) was that `repartitionByRange` on a set of strings doesn't know about relativity/hirearchy. As you point out, this isn't really "sorting", but "clustering". 

From R3: 

> Changed the clustering column to be a struct referencing the actual columns instead of a contacted string p1::p2::p3.
> Thanks to @ Nandhakumar Saravana for pointing this out, when clustering by a string a::b::c spark doesn't know about relative priority/hierarchy. AKA a::b::c should be close to a::b::d, which it does when clustering by actual columns and not a String literal.

What happens when you use a single String like this is that you don't get good clustering on the keys in the order that you specify, unless you happen to have > 1 manifest files' worth of manifests to write. 

In practice, this means that a table that is initially defined as `PARTITIONED BY (a, b, c)` will have much better metadata organization than one `PARTITIONED BY (c, b, a)` with `RewriteManifests.sort(a, b, c)`. Which is the behavior I'm shoting for here.

zachdisc commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -628,4 +613,50 @@ private Table table() {
     }
   }
+
+  static class PartitionFieldSortFunction implements SerializableFunction<DataFile, String> {
```
You left a comment before about including the function as a static class in here, this might be the trick to get it to work for me.

zachdisc commented:
@nastra nastra 

> @zachdisc Hey Zach, sorry I was trying to experiment with refactoring the current PR, but somehow the branch is public and I ended up directly committing to it...
> 
> Let me know if you agree with this or not, overall it tries to express the partition column sort order approach as a kind of sort function to minimize the if-else branches.

I left a comment in the commit - I like the cleanliness of what you proposed but I don't think it yields quite as performant a result unfortunately.

zachdisc commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -309,24 +312,6 @@ private List<ManifestFile> writePartitionedManifests(
-      Column[] actualPartitionColumns =
-          partitionFieldSortOrder.stream()
-              .map(p -> col("data_file.partition." + p))
-              .toArray(Column[]::new);
```
This part might be the trick though, I wonder if we can modify the `PartitionSortFunction` to return a column, which we could define as a struct with partition column names in the standard `sort` case, and just a String in the more customizable case.

Actually, I'm wondering if we could make it return a struct in general, or something that we could treat as such. For the same reason as I give below with wanting to deliver some level of hierarchy.

jackye1995 commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +47,39 @@ public interface RewriteManifests
+   * Rewrite manifests in a given order, based on partition field names
+   *
+   * <p>Supply an optional set of partition field names to sort the rewritten manifests by. Expects
+   * exact transformed column names used for partitioning; not the raw columnnames that partitions
+   * are derived from. E.G. supply 'data_bucket' and not 'data' for a bucket(N, data) partition
+   * definition
```
I think L54-56 can be moved to L61 as a part of describing the specific parameter.

jackye1995 commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +47,39 @@ public interface RewriteManifests
+   * @param partitionFieldSortOrder a list of partition field names
+   * @return this for method chaining
+   */
+  default RewriteManifests sort(List<String> partitionFieldSortOrder) {
```
One thing to double click here with everyone. I think we originally landed the API name as `sort`, but the implementation is using `partitionByRange`. In that case, I feel `clusterBy` is a more suitable name for this API (and also the other one below). What do we think? @zachdisc @RussellSpitzer @nastra

zachdisc replied:
I agree and vote for `clusterBy`. Sounds like we have two votes, I'll refactor unless I hear strong opposition :-)

jackye1995 commented:
>  I don't think it yields quite as performant a result unfortunately.

I see that's a good point, using the UDF function definitely is less performant than sorting native columns, in that case we can keep it as is, unless anyone has a way to fix the perf.

zachdisc commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +47,39 @@ public interface RewriteManifests
   RewriteManifests rewriteIf(Predicate<ManifestFile> predicate);
 
+  /**
+   * Rewrite manifests in a given order, based on partition field names
```
Will do

zachdisc commented:
> I see that's a good point, using the UDF function definitely is less performant than sorting native columns, in that case we can keep it as is, unless anyone has a way to fix the perf.

I'll see if I can try to return a struct column which would match performance - at which point the API wouldn't change it would just be a refactor/simplification within the RewriteManifestsAction itself though

zachdisc commented:
#### R4
Renamed `sort` to `clusterBy` - more honest and aligns with the core API as well. 

Have not made progress on using the custom UDF for both standard partition column ordering as well as for custom functionality. But that is an internal implementation detail that I think could be addressed later if we're okay with that.

zachdisc commented:
@jackye1995 @nastra any comments or thoughts on this?

ZachDischner commented:
Any interest in this? I've been using it in the wild and this is a nice feature!

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +47,43 @@ public interface RewriteManifests
+  /**
+   * Rewrite manifests in a given order, based on partition field names
+   *
+   * <p>Supply an optional set of partition field names to cluster the rewritten manifests by. For
```
* Avoid "you"

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +47,43 @@ public interface RewriteManifests
+   * <p>If not set, manifests will be rewritten in the order of the transforms in the table's
+   * current partition spec.
+   *
+   * @param partitionFieldClustering Exact transformed column names used for partitioning; not the
```
partitionTransform?

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +47,43 @@ public interface RewriteManifests
+   *     clustering
+   * @return this method for chaining
+   */
+  default RewriteManifests clusterBy(Function<DataFile, String> clusterStrategyFunction) {
```
I don't feel good about opening up a public api like this that takes an arbitrary function unless we have a very good reason to. Do we need this kind of flexability?

ZachDischner replied:
I can say that there are reasons to, but it is up to you to decide if they are very good reasons. 

For example, I know that a primary use case for `bucket(someId, 10000)` partitioned tables will require reading `bucket_someId=1` and `bucket_someId=999` data together, so I'll want to cluster my planning around this use case. 

Potentially simpler, say I'm partitioning by `month(timestamp)`. I can take the min/max values of a data file and cluster the files in a given month partition by day for more efficient query planning. This option lets power users take what Iceberg gives you to a new level.

ZachDischner replied:
I should also mention that this is a feature in the public Java API, it just doesn't scale for large tables which is why I wanted to implement a Spark version 

https://github.com/apache/iceberg/blob/5e279c868f8e2087f88ae779d8cc8474768bc5e2/api/src/main/java/org/apache/iceberg/RewriteManifests.java#L49

RussellSpitzer replied:
@zachdisc I really don't want to widen the api if we can at all help it, adding a Function<> api is something that will be hard for us to walk back. I would feel more comfortable if we started with partition_transforms and consider opening things up more as community demand becomes a bit more pronounced. 

I think a lot of your use cases will work better if we allow users to sort on any column. We can't really do that at the moment, but we have a lot of plans for V4 metadata that will allow us to bring all column min/maxes up to the manifest_list level.

ZachDischner replied:
Thanks @RussellSpitzer. Since it is just you and I here I'll happily disagree and commit. I'll get rid of arbitrary function based reordering.

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -160,6 +172,32 @@ public RewriteManifestsSparkAction stagingLocation(String newStagingLocation) {
   }
 
+  @Override
+  public RewriteManifestsSparkAction clusterBy(List<String> clusteringColumns) {
```
clusteringColumns -> partitionFields?

ZachDischner replied:
Can change

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -44,6 +47,43 @@ public interface RewriteManifests
+   *     for a bucket(N, data) partition * definition
+   * @return this for method chaining
+   */
+  default RewriteManifests clusterBy(List<String> partitionFieldClustering) {
```
List<String> partitionFields

ZachDischner replied:
Can change

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -160,6 +172,32 @@ public RewriteManifestsSparkAction stagingLocation(String newStagingLocation) {
+        spec.fields().stream().map(PartitionField::name).collect(Collectors.toSet());
+
+    // Check if these partition fields are included in the spec
+    Preconditions.checkArgument(
```
Do we care about case sensitivity here?

ZachDischner replied:
I believe partition columns are saved verbatim. Are `FOO` and `foo` considered the same or different columns? We could desensitize this operation to case

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -160,6 +172,32 @@ public RewriteManifestsSparkAction stagingLocation(String newStagingLocation) {
+        "Cannot use custom clustering to rewrite manifests '%s'. All partition columns must be "
+            + "defined in the current partition spec: %s. Choose from the available partitionable columns: %s",
+        clusteringColumns,
+        this.spec.specId(),
```
Why should the spec ID and not the spec?

I think this should probably be something like
```
"Cannot set manifest clustering because specified  field(s) %s were not found in current partition spec %s.",
missingFields,
this.spec
```

ZachDischner replied:
Can change

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteManifestsSparkAction.java:
```diff
@@ -250,12 +288,59 @@ private List<ManifestFile> writeUnpartitionedManifests(
+          manifestEntryDF.withColumn(
+              CUSTOM_CLUSTERING_COLUMN_NAME, functions.struct(actualPartitionColumns));
+    } else {
+      clusteredManifestEntryDF =
```
Rather than having a third branch we should probably just have `partitionFieldClustering` set to data_file.partition by default somehow. Just trying to see if we have a way to remove the complexity here.

ZachDischner replied:
I'll give it a shot.

RussellSpitzer commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +474,310 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+
+    SparkActions actions = SparkActions.get();
+
+    //  c2 is not a partition column, cannot use for clustering
```
This comment probably is not required

RussellSpitzer commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +474,310 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+    SparkActions actions = SparkActions.get();
+
+    //  c2 is not a partition column, cannot use for clustering
+    List<String> badClusteringKeys1 = ImmutableList.of("c1", "c2");
```
I would call this "nonExistentField"

RussellSpitzer commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java:
```diff
@@ -466,6 +474,310 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+
+    // c3_bucket is the correct internal partition name to use, c3 is the untransformed column name,
+    // clusterBy() expects the hidden partition column names
+    List<String> badClusteringKeys2 = ImmutableList.of("c1", "c3");
```
wrongFieldName

RussellSpitzer replied:
In general we shouldn't use variables with a numerical suffix if we can give them a more meaningful name

ZachDischner commented:
I'm going to work on this

zachdisc commented:
I rebased locally and resolved merge conflicts and addressed Russel's ask to remove the function-based rewrite sorting.

zachdisc commented:
Please let me know the best way to deal with merge conflicts. I thought to rebase and get everything back in sync with the main branch, but that looks like the wrong flow here. I can close this PR and reopen a fresh one if that's preferred.

zachdisc commented:
Created a fresh PR with feedback here incorporated - https://github.com/apache/iceberg/pull/11881

This got messy with the accumulated changes and conflicts (paired with my likely messing things up with a rebase). Suggest closing this out in favor of the fresh start above.

zachdisc commented:
Closing in favor of https://github.com/apache/iceberg/pull/11881
